// graphql-yoga/esm/plugins/use-unhandled-route.js
import { isPromise } from '@graphql-tools/utils';
import landingPageBody from '../landing-page-html.js';
export const defaultRenderLandingPage = function defaultRenderLandingPage(opts) {
    return new opts.fetchAPI.Response(landingPageBody
        .replace(/__GRAPHIQL_LINK__/g, opts.graphqlEndpoint)
        .replace(/__REQUEST_PATH__/g, opts.url.pathname), {
        status: 200,
        statusText: 'OK',
        headers: {
            'Content-Type': 'text/html',
        },
    });
};
export function useUnhandledRoute(args) {
    const landingPageRenderer = args.landingPageRenderer || defaultRenderLandingPage;
    return {
        onRequest({ request, fetchAPI, endResponse, url }) {
            const graphqlEndpoint = args.getGraphQLEndpoint();
            if (!request.url.endsWith(graphqlEndpoint) &&
                !request.url.endsWith(`${graphqlEndpoint}/`) &&
                url.pathname !== graphqlEndpoint &&
                url.pathname !== `${graphqlEndpoint}/` &&
                !args.getGraphQLEndpointURLPattern().test(url)) {
                if (args.showLandingPage === true &&
                    request.method === 'GET' &&
                    !!request.headers?.get('accept')?.includes('text/html')) {
                    const landingPage$ = landingPageRenderer({
                        request,
                        fetchAPI,
                        url,
                        graphqlEndpoint,
                        get urlPattern() {
                            return args.getGraphQLEndpointURLPattern();
                        },
                    });
                    if (isPromise(landingPage$)) {
                        return landingPage$.then(endResponse);
                    }
                    endResponse(landingPage$);
                    return;
                }
                endResponse(new fetchAPI.Response('', {
                    status: 404,
                    statusText: 'Not Found',
                }));
            }
        },
    };
}
